compiler/rustc_lint/src/let_underscore.rs

```diff
@@ -0,0 +1,270 @@
+        fn is_must_use_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+            match ty.kind() {
+                ty::Adt(adt, _) => has_must_use_attr(cx, adt.did()),
+                ty::Foreign(ref did) => has_must_use_attr(cx, *did),
+                ty::Slice(ty)
+                | ty::Array(ty, _)
+                | ty::RawPtr(ty::TypeAndMut { ty, .. })
+                | ty::Ref(_, ty, _) => {
+                    // for the Array case we don't need to care for the len == 0 case
+                    // because we don't want to lint functions returning empty arrays
+                    is_must_use_ty(cx, *ty)
+                }
+                ty::Tuple(substs) => substs.iter().any(|ty| is_must_use_ty(cx, ty)),
+                ty::Opaque(ref def_id, _) => {
+                    for (predicate, _) in cx.tcx.explicit_item_bounds(*def_id) {
+                        if let ty::PredicateKind::Trait(trait_predicate) =
+                            predicate.kind().skip_binder()
+                        {
+                            if has_must_use_attr(cx, trait_predicate.trait_ref.def_id) {
+                                return true;
+                            }
+                        }
+                    }
+                    false
+                }
+                ty::Dynamic(binder, _) => {
+                    for predicate in binder.iter() {
+                        if let ty::ExistentialPredicate::Trait(ref trait_ref) =
+                            predicate.skip_binder()
+                        {
+                            if has_must_use_attr(cx, trait_ref.def_id) {
+                                return true;
+                            }
+                        }
+                    }
+                    false
+                }
+                _ => false,
+            }
+        }
+
+        // check if expr is calling method or function with #[must_use] attribute
+        fn is_must_use_func_call(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+            let did = match expr.kind {
+                hir::ExprKind::Call(path, _) if let hir::ExprKind::Path(ref qpath) = path.kind => {
+                    if let hir::def::Res::Def(_, did) = cx.qpath_res(qpath, path.hir_id) {
+                        Some(did)
+                    } else {
+                        None
+                    }
+                },
+                hir::ExprKind::MethodCall(..) => {
+                    cx.typeck_results().type_dependent_def_id(expr.hir_id)
+                }
+                _ => None,
+            };
+
+            did.map_or(false, |did| has_must_use_attr(cx, did))
+        }
+
+        // returns true if DefId contains a `#[must_use]` attribute
+        fn has_must_use_attr(cx: &LateContext<'_>, did: hir::def_id::DefId) -> bool {
+            cx.tcx
+                .get_attrs(did, rustc_span::sym::must_use)
+                .find(|a| a.has_name(rustc_span::sym::must_use))
```

Comment:
Isn't this `has_name` call trivially true?
IIUC, this construct only tests that the return of `get_attrs` is not empty.
Can `cx.tcx.has_attr(did, rustc_span::sym::must_use)` be used?